src/ports/postgres/modules/graph/pagerank.py_in

```diff
@@ -0,0 +1,290 @@
+                                  """)[0]['setting']
+    plpy.execute('SET client_min_messages TO warning')
+    params_types = {'src': str, 'dest': str}
+    default_args = {'src': 'src', 'dest': 'dest'}
+    edge_params = extract_keyvalue_params(edge_args, params_types, default_args)
+
+    # populate default values for optional params if null
+    if damping_factor is None:
+        damping_factor = 0.85
+    if max_iter is None:
+        max_iter = 100
+    if threshold is None:
+        threshold = 0.00001
+    if vertex_id is None:
+        vertex_id = "id"
+    validate_pagerank_args(vertex_table, vertex_id, edge_table, edge_params,
+        out_table, damping_factor, max_iter, threshold, 'PageRank')
+    src = edge_params["src"]
+    dest = edge_params["dest"]
+
+    edge_temp_table = unique_string(desp='temp_edge')
+    distribution = m4_ifdef(<!__POSTGRESQL__!>, <!''!>,
+        <!"DISTRIBUTED BY ({0})".format(dest)!>)
+    plpy.execute("""
+        DROP TABLE IF EXISTS {edge_temp_table};
+        CREATE TEMP TABLE {edge_temp_table} AS
+        SELECT * FROM {edge_table}
+        {distribution}
+        """.format(**locals()))
+    # GPDB and HAWQ have distributed by clauses to help them with indexing.
+    # For Postgres we add the indices manually.
+    sql_index = m4_ifdef(<!__POSTGRESQL__!>,
+        <!"""CREATE INDEX ON {edge_temp_table} ({src});
+        """.format(**locals())!>,
+        <!''!>)
+    plpy.execute(sql_index)
+
+    nvertices = plpy.execute("""
+            SELECT COUNT({0}) AS cnt
+            FROM {1}
+        """.format(vertex_id, vertex_table))[0]["cnt"]
+    init_value = 1.0/nvertices
+    random_prob = (1.0-damping_factor)/nvertices
+    cur = unique_string(desp='cur')
+    message = unique_string(desp='message')
+    plpy.execute("""
+            CREATE TEMP TABLE {cur} AS
+            SELECT {vertex_id}, {init_value}::DOUBLE PRECISION AS pagerank
+            FROM {vertex_table}
+        """.format(**locals()))
+    v1 = unique_string(desp='v1')
+
+    out_cnts = unique_string(desp='out_cnts')
+    out_cnts_cnt = unique_string(desp='cnt')
+    # Compute the out-degree of every node in the graph.
+    cnts_distribution = m4_ifdef(<!__POSTGRESQL__!>, <!''!>,
+        <!"DISTRIBUTED BY ({0})".format(vertex_id)!>)
+
+    plpy.execute("""
+        DROP TABLE IF EXISTS {out_cnts};
+        CREATE TEMP TABLE {out_cnts} AS
+        SELECT {src} AS {vertex_id}, COUNT({dest}) AS {out_cnts_cnt}
+        FROM {edge_table}
+        GROUP BY {src}
+        {cnts_distribution}
+        """.format(**locals()))
+
+    for i in range(max_iter):
+        plpy.execute("""
+                CREATE TABLE {message} AS
+                SELECT {edge_temp_table}.{dest} AS {vertex_id},
+                        SUM({v1}.pagerank/{out_cnts}.{out_cnts_cnt})*{damping_factor}+{random_prob} AS pagerank
+                FROM {cur}, {out_cnts}, {edge_temp_table}, {cur} AS {v1}
```

Comment:
Will use joins instead of this, but not sure of the performance gain yet. Might have to revert if the results are not favorable.